doc/dev/sharp_bits.rst

```diff
@@ -0,0 +1,1015 @@
+
+.. raw:: html
+
+    <div class="summary-table">
+
+.. autosummary::
+    :nosignatures:
+
+    ~catalyst.debug.print
+    ~catalyst.autograph_source
+
+.. raw:: html
+
+    </div>
+    <div class="summary-table">
+
+
+Compile-time vs. runtime
+------------------------
+
+An important distinction to make in Catalyst, which we typically don't have to
+worry about with standard PennyLane, is the concept of **compile time**
+vs. **runtime**.
+
+Very roughly, the following three processes occur when using the :func:`@qjit <.qjit>` decorator
+with just-in-time (JIT) compilation.
+
+#. **Program capture or tracing:** When the :func:`@qjit <.qjit>` decorated function is
+   first called (or, when the :func:`@qjit <.qjit>` is first applied if using function
+   type hints and :ref:`ahead-of-time mode <ahead_of_time>`), Catalyst
+   will 'capture' the entire hybrid workflow with **placeholder variables of
+   unknown value** used as the function arguments
+   (the **runtime arguments**).
+
+   These symbolic tracer objects represent **dynamic variable**, and are used
+   to determine how the JIT compiled function transforms its inputs to
+   outputs.
+
+#. **Compilation:** The captured program is then compiled to a parametrized
+   binary using the Catalyst compiler.
+
+#. **Execution:** Finally, the compiled function is executed with the
+   provided numerical function inputs, and the results returned.
+
+Once the function is first compiled, subsequent executions of the function
+will simply re-use the previous compiled binary, allowing steps (1) and (2) to
+be skipped. (Note: some cases, such as if the function argument types change,
+may trigger re-compilation.)
+
+For example, consider the following, where we print out a variable in the middle of
+our :func:`@qjit <~.qjit>` compiled function:
+
+>>> @qjit
+... def f(x):
+...     print(f"x = {x}")
+...     return x ** 2
+>>> f(2.)
+x = Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
+array(4.)
+>>> f(3.)
+array(9.)
+
+We can see that on the first execution, program capture/tracing occurs, and we
+can see the dynamic variable is printed (tracers capture *type*
+and *shape*, but not numeric value). This captured program is compiled, and
+then the binary is executed directly to return the function value --- the
+print statement is never invoked with the numerical value of ``x``.
+
+When we execute the function again, steps (1) and (2) are skipped since we
+have already compiled a binary; this is called directly to get the function
+result, and again the print statement is never hit.
+
+This allows us to distinguish between computations that happen
+at **compile-time** (steps 1 and 2), such as the ``print`` statement above,
+and those that happen at **runtime** (step 3).
+
+.. note::
+
+    As a general rule of thumb, for a function that is repeatedly executed
+    with different parameters, we want as much evaluation as possible to
+    happen at compile time.
```

Comment:
I think the JAX docs also say this somewhere :) @dime10 do you want any changes here?

Reply:
Do you know where it says that?

One reason JAX introduced omnistaging is so that expensive operations on constants are still staged out to accelerators. So I'm not quite sure what to make of your statement.

Reply:
Oh good point. I can delete it 👍 

Before I do though, is there any (very general) rule of thumb you would like to have people keep in their mind while writing catalyst programs?

Reply:
Hmm, in the context of this paragraph, maybe that generally things that happen at compile-time are slow (or lead to slowdowns), while things that happen at runtime are fast (or lead to speadups). (Think compile-time vs runtime loops, computing in Python vs in native code).

However, if the same computation is repeated every time the compiled function is run (where the results are the same no matter the inputs), and it is expensive, then it may be worth doing the computation once in Python and use the results statically in the program (for example Hamiltonian/chemistry calculations?).

Reply:
```suggestion
    As a general rule of thumb, things that happen at compile-time
    are slow (or lead to slowdowns), while things that happen at
    runtime are fast (or lead to speadups).

    However, if the same computation is repeated every time the
    compiled function is run (where the results are the same no
    matter the inputs), and it is expensive, then it may be worth
    doing the computation once in Python and use the results
    statically in the program.
```

Perfect 💯